dev/PushNotifications/PushNotificationReceivedEventArgs.cpp

```diff
@@ -0,0 +1,118 @@
+﻿#include "pch.h"
+
+#include <winrt/Windows.ApplicationModel.background.h>
+#include <winrt/Windows.Storage.Streams.h>
+#include <winrt/Windows.Networking.PushNotifications.h>
+
+#include "PushNotificationReceivedEventArgs.h"
+#include "Microsoft.Windows.PushNotifications.PushNotificationReceivedEventArgs.g.cpp"
+#include <iostream>
+
+namespace winrt
+{
+    using namespace Windows::ApplicationModel::Background;
+    using namespace Windows::Storage::Streams;
+    using namespace Windows::Networking::PushNotifications;
+}
+
+namespace winrt::Microsoft::Windows::PushNotifications::implementation
+{
+    PushNotificationReceivedEventArgs::PushNotificationReceivedEventArgs(winrt::IBackgroundTaskInstance const& backgroundTask)
+    {
+        m_backgroundTaskInstance = backgroundTask;
+
+        // Need to process the RawNotification payload here because TriggerDetails is set to nullptr leaving IBackgroundTaskInstance::Run 
+        auto triggerDetails = m_backgroundTaskInstance.TriggerDetails();
+        RawNotification rawNotification = triggerDetails.as<RawNotification>();
+
+        IBuffer rawContentAsBuffer = rawNotification.ContentBytes();
+        DataReader dataReader = DataReader::FromBuffer(rawContentAsBuffer);
+
+        m_backgroundPayload = winrt::com_array<uint8_t>(rawContentAsBuffer.Length());
+        dataReader.ReadBytes(m_backgroundPayload);
```

Comment:
Working way too hard.

```suggestion
        m_backgroundPayload = { rawContentAsBuffer.data(), rawContentAsBuffer.Length() };
```

And then we find that this code is repeated later for the `m_args` case, so we may as well factor it out.

```cpp
winrt::com_array<uint8_t> GetRawPayload(RawNotification const& rawNotification)
{
    IBuffer rawContentAsBuffer = rawNotification.ContentBytes();
    return { rawContentAsBuffer.data(), rawContentAsBuffer.Length() };
}
```

Now you can use it to construct the member.

```cpp
    m_backgroundPayload{ GetRawPayload(backgroundTask.TriggerDetails().as<RawNotification>()) }
    {
    }
```

This allows you to make all the members const.

Reply:
IBuffer doesn't have a data function so we have to to use the data reader. I will create function that does the RawPayload processing and allows us to use the intializer_list constructor.

Reply:
IBuffer does have a `data()` member function. It's a bonus feature of C++/WinRT.

```cpp
    template <typename D>
    struct consume_Windows_Storage_Streams_IBuffer
    {
        [[nodiscard]] WINRT_IMPL_AUTO(uint32_t) Capacity() const;
        [[nodiscard]] WINRT_IMPL_AUTO(uint32_t) Length() const;
        WINRT_IMPL_AUTO(void) Length(uint32_t value) const;

        auto data() const
        {
            uint8_t* data{};
            static_cast<D const&>(*this).template as<IBufferByteAccess>()->Buffer(&data);
            return data;
        }
    };
```

[We use it in the 3D Printing sample](https://github.com/microsoft/Windows-universal-samples/blob/b1cb20f191d3fd99ce89df50c5b7d1a6e2382c01/Samples/3DPrinting/cppwinrt/Scenario1_Print.cpp#L33), for example.

Reply:
I've tried implementing the above suggestion but I am getting a compiler error: cannot convert from 'initializer list' to 'winrt::com_array<uint8_t>'. 

What am I missing here?

Reply:
Sorry, I confused `com_array` with `array_view`. You can look at the available constructors of `winrt::com_array` to see what to do. (The two-iterator form seems to be appropriate here.)

Reply:
Do we even have to return a `com_array` at all? Could we just return the `IBuffer`?

Reply:
Here is the current implementation I am using:
`return { rawContentAsBuffer.data(), rawContentAsBuffer.data() + (rawContentAsBuffer.Length() * sizeof(uint8_t)) };`

We could store the IBuffer in the constructor and return a com_array on calls to Payload() as listed in the spec. What do you think?